Sources/FoundationEssentials/Subprocess/Platforms/Subprocess+Darwin.swift

```diff
@@ -0,0 +1,232 @@
+        // Set QualityOfService
+        // spanattr_qos seems to only accept `QOS_CLASS_UTILITY` or `QOS_CLASS_BACKGROUND`
+        // and returns an error of `EINVAL` if anything else is provided
+        if spawnAttributeError == 0 && self.platformOptions.qualityOfService == .utility{
+            spawnAttributeError = posix_spawnattr_set_qos_class_np(&spawnAttributes, QOS_CLASS_UTILITY)
+        } else if spawnAttributeError == 0 && self.platformOptions.qualityOfService == .background {
+            spawnAttributeError = posix_spawnattr_set_qos_class_np(&spawnAttributes, QOS_CLASS_BACKGROUND)
+        }
+
+        // Setup cwd
+        var chdirError: Int32 = 0
+        if intendedWorkingDir != .currentWorkingDirectory {
+            chdirError = intendedWorkingDir.withPlatformString { workDir in
+                return posix_spawn_file_actions_addchdir_np(&fileActions, workDir)
+            }
+        }
+
+        // Error handling
+        if chdirError != 0 || spawnAttributeError != 0 {
+            try self.cleanupAll(input: input, output: output, error: error)
+            if spawnAttributeError != 0 {
+                throw POSIXError(.init(rawValue: result) ?? .ENODEV)
+            }
+
+            if chdirError != 0 {
+                throw CocoaError(.fileNoSuchFile, userInfo: [
+                    .debugDescriptionErrorKey: "Cannot failed to change the working directory to \(intendedWorkingDir) with errno \(chdirError)"
+                ])
+            }
+        }
+        // Run additional config
+        if let spawnConfig = self.platformOptions.additionalSpawnAttributeConfigurator {
+            try spawnConfig(&spawnAttributes)
+        }
+        if let fileAttributeConfig = self.platformOptions.additionalFileAttributeConfigurator {
+            try fileAttributeConfig(&fileActions)
+        }
+        // Spawn
+        var pid: pid_t = 0
+        let spawnError: CInt = executablePath.withCString { exePath in
+            return supplementaryGroups.withOptionalUnsafeBufferPointer { sgroups in
+                return _subprocess_spawn(
+                    &pid, exePath,
+                    &fileActions, &spawnAttributes,
+                    argv, env,
+                    uidPtr, gidPtr,
+                    Int32(supplementaryGroups?.count ?? 0), sgroups?.baseAddress,
+                    self.platformOptions.createSession ? 1 : 0
+                )
+            }
+        }
+        // Spawn error
+        if spawnError != 0 {
+            try self.cleanupAll(input: input, output: output, error: error)
+            throw POSIXError(.init(rawValue: spawnError) ?? .ENODEV)
+        }
+        return Subprocess(
+            processIdentifier: .init(value: pid),
+            executionInput: input,
+            executionOutput: output,
+            executionError: error
+        )
+    }
+}
+
+// Special keys used in Error's user dictionary
+extension String {
+    static let debugDescriptionErrorKey = "NSDebugDescription"
```

Comment:
We have a pattern of defining keys like this at the bottom of `CocoaError.swift` instead of defining them as extensions on `String`, should we follow that same pattern for this one?